src/probnum/random_variables/_normal.py

```diff
@@ -174,29 +182,30 @@ def __init__(
             var = self._dense_var
             entropy = self._dense_entropy
 
-            if cov_cholesky is None:
-                self._compute_cov_cholesky = self.dense_cov_cholesky
-            else:
-                if not isinstance(cov_cholesky, type(self._cov)):
+            self._compute_cov_cholesky = self.dense_cov_cholesky
+
+            # Ensure that the Cholesky factor has the same type as the covariance,
+            # and, if necessary, promote data types. Check for (in this order): type, shape, dtype.
+            if self._cov_cholesky is not None:
+
+                if not isinstance(self._cov_cholesky, type(self._cov)):
                     raise ValueError(
                         f"The covariance matrix is of type `{type(self._cov)}`, so its "
                         f"Cholesky decomposition must be of the same type, but an "
-                        f"object of type `{type(cov_cholesky)}` was given."
+                        f"object of type `{type(self._cov_cholesky)}` was given."
                     )
 
-                if cov_cholesky.shape != self._cov.shape:
+                if self._cov_cholesky.shape != self._cov.shape:
                     raise ValueError(
                         f"The cholesky decomposition of the covariance matrix must "
                         f"have the same shape as the covariance matrix, i.e. "
-                        f"{self._cov.shape}, but shape {cov_cholesky.shape} was given"
+                        f"{self._cov.shape}, but shape {self._cov_cholesky.shape} was given"
                     )
 
-                if cov_cholesky.dtype != self._cov.dtype:
+                if self._cov_cholesky.dtype != self._cov.dtype:
                     # TODO: Implement casting for linear operators
-                    if not isinstance(cov_cholesky, linops.LinearOperator):
-                        cov_cholesky = cov_cholesky.astype(self._cov.dtype)
-
-                self._compute_cov_cholesky = lambda: cov_cholesky
+                    if not isinstance(self._cov_cholesky, linops.LinearOperator):
+                        self._cov_cholesky = self._cov_cholesky.astype(self._cov.dtype)
```

Comment:
I am not sure how robust it is to always promote the Cholesky data type to that of the covariance. For instance, I'd think it is perfectly okay to specify an integer variance (say 2), which has a non-integer Cholesky (sqrt(2)). If this is promoted, information is lost. 

In any case, I did not feel confident enough to change things here, so instead I included this promotion of types into the docstring (in case people use it and are wondering where information is lost).

Reply:
In the long run, we should probably restrict users from using non-float dtypes here, but for now I'd say it's fine.

Reply:
I agree, but the same conversation might come up in terms of `float32` and `float64`. Though, this is neither relevant for this PR nor sufficiently urgent to raise an issue (especially since its documented fairly prominently now).